docs/design/2024-09-06-maintain-priority-queue-in-memory.md

```diff
@@ -0,0 +1,557 @@
+
+1. We start a background worker to call the statsHandle.Update(do.InfoSchema()) function every 3 seconds(stats lease).
+2. In the stats cache struct, we maintain an atomic timestamp called: maxTblStatsVer
+3. Whenever we call the statsHandle.Update(do.InfoSchema()) function, we query the new update from incrementally updating.
+
+    ```sql
+    SELECT version, table_id, modify_count, count from mysql.stats_meta where version > %? order by version
+    ```
+
+4. Once we have all the tables with new updates, we update the statistics cache for them. At the same time, we will update maxTblStatsVer as well.
+
+This way, we don't have to scan the whole mysql.stats_meta, but we can still keep the latest meta information in memory.
+We can use the same approach for the priority queue. Every time we find an update from the stats meta table, we know that we should recalculate the score for the table. Because the modify_count and count might have changed.
+
+#### DDL
+
+Similarly, we can apply this mechanism to DDL events. We need an API to query DDLs that happened in the past, and the query interface should be very similar to the way you query for data changes that happened.
+Right now, we only have an in-memory channel to send the DDL events to the statistics module from the DDL module. So we need to find a reliable way to store and fetch DDL events.
+
+To ensure that DDL events are processed correctly, DDL module writes them to a physical table with preset primary key in the same transaction of finishing DDL. The stats module reads these events sorted by their primary key, leveraging the fact that DDL events are executed in order, which guarantees that their IDs are also ordered.
+We use a large transaction to process the DDL events within the stats module and delete them once processed. This approach eliminates the need for specific checkpoints during failover and guarantees exactly once delivery, as each event is read and processed only once.
+
+### The New Priority Queue - Maintain the heap in memory
+
+Because we are constantly adding, removing, and updating the heap, we need to use a thread-safe heap to store analysis jobs.
+The required interfaces look like the following:
+
+| Operation | Time complexity |
+| --------- | --------------- |
+| Add       | O(log n)        |
+| Update    | O(log n)        |
+| Delete    | O(log n)        |
+| Pop       | O(1)            |
+| Get       | O(1)            |
+
+#### Information to be stored
+
+```go
+type Indicators struct {
+        // ChangePercentage is the percentage of the changed rows.
+        // Usually, the more the changed rows, the higher the priority.
+        // It is calculated by modifiedCount / last time analysis count.
+        ChangePercentage float64
+        // TableSize is the table size in rows * len(columns).
+        TableSize float64
+        // LastAnalysisDuration is the duration from the last analysis to now.
+        // In seconds.
```

Comment:
"In seconds" means 1 second should be store as `time.Duration(1)` rather than `1 * time.Second` which is `1000000000`? Easy to make mistake.

Reply:
Oh, good point. It shouldn't say "In seconds" here. In code, we actually use nanoseconds.

Reply:
Updated.

Reply:
I will not remove the code comment in this PR. This PR is only for the design doc.

Reply:
I will fix it in another PR.